File: web/frontend/src/utils/errorHandler.ts
import { App } from 'vue'
import { api } from '@/api'

interface ErrorResponse {
  success: false
  error: {
    code: string
    message: string
    status_code: number
  }
}

interface ApiError extends Error {
  response?: {
    data: ErrorResponse
    status: number
  }
}

export class APIError extends Error {
  code: string
  status: number
  message: string

  constructor(message: string, code: string = 'API_ERROR', status: number = 500) {
    super(message)
    this.name = 'APIError'
    this.code = code
    this.status = status
    this.message = message
  }
}

export class ValidationError extends APIError {
  constructor(message: string) {
    super(message, 'VALIDATION_ERROR', 400)
  }
}

export class NotFoundError extends APIError {
  constructor(message: string = '资源未找到') {
    super(message, 'NOT_FOUND', 404)
  }
}

export class UnauthorizedError extends APIError {
  constructor(message: string = '未授权') {
    super(message, 'UNAUTHORIZED', 401)
  }
}

export class ForbiddenError extends APIError {
  constructor(message: string = '禁止访问') {
    super(message, 'FORBIDDEN', 403)
  }
}

export class RateLimitError extends APIError {
  constructor(message: string = '请求过于频繁') {
    super(message, 'RATE_LIMIT_EXCEEDED', 429)
  }
}

export class ServiceUnavailableError extends APIError {
  constructor(message: string = '服务暂时不可用') {
    super(message, 'SERVICE_UNAVAILABLE', 503)
  }
}

export function handleApiError(error: any): APIError {
  if (error instanceof APIError) {
    return error
  }

  if (error.response) {
    const response = error.response
    
    if (response.data && response.data.error) {
      const errorData = response.data.error
      
      switch (errorData.code) {
        case 'VALIDATION_ERROR':
          return new ValidationError(errorData.message)
        case 'NOT_FOUND':
          return new NotFoundError(errorData.message)
        case 'UNAUTHORIZED':
          return new UnauthorizedError(errorData.message)
        case 'FORBIDDEN':
          return new ForbiddenError(errorData.message)
        case 'RATE_LIMIT_EXCEEDED':
          return new RateLimitError(errorData.message)
        case 'SERVICE_UNAVAILABLE':
          return new ServiceUnavailableError(errorData.message)
        default:
          return new APIError(errorData.message, errorData.code, errorData.status_code)
      }
    }
    
    // 没有标准错误格式，使用状态码判断
    switch (response.status) {
      case 400:
        return new ValidationError('请求参数错误')
      case 401:
        return new UnauthorizedError()
      case 403:
        return new ForbiddenError()
      case 404:
        return new NotFoundError()
      case 429:
        return new RateLimitError()
      case 503:
        return new ServiceUnavailableError()
      default:
        return new APIError('服务器错误', 'INTERNAL_ERROR', response.status)
    }
  }

  if (error.request) {
    // 请求已发送但没有收到响应
    return new ServiceUnavailableError('网络连接失败，请检查网络设置')
  }

  // 其他错误
  return new APIError(error.message || '未知错误', 'UNKNOWN_ERROR', 500)
}

export function setupGlobalErrorHandler(app: App) {
  app.config.errorHandler = (error, instance, info) => {
    console.error('全局错误:', error)
    console.error('错误信息:', info)
    
    // 可以在这里添加错误上报逻辑
    // 例如：发送到错误监控服务
  }

  window.addEventListener('unhandledrejection', (event) => {
    console.error('未处理的Promise拒绝:', event.reason)
    
    // 可以在这里添加错误上报逻辑
  })

  window.addEventListener('error', (event) => {
    console.error('全局错误:', event.error)
    
    // 可以在这里添加错误上报逻辑
  })
}

export function showErrorToast(error: APIError | Error, duration: number = 3000) {
  // 这里可以集成Element Plus的ElMessage或其他通知组件
  console.error('错误:', error.message)
  
  // 简单实现，实际项目中可以使用更好的通知组件
  const toast = document.createElement('div')
  toast.className = 'error-toast'
  toast.textContent = error.message
  toast.style.cssText = `
    position: fixed;
    top: 20px;
    right: 20px;
    padding: 12px 20px;
    background: #f56c6c;
    color: white;
    border-radius: 4px;
    box-shadow: 0 2px 12px rgba(0, 0, 0, 0.15);
    z-index: 9999;
    animation: slideIn 0.3s ease;
  `
  
  document.body.appendChild(toast)
  
  setTimeout(() => {
    toast.style.animation = 'slideOut 0.3s ease'
    setTimeout(() => {
      document.body.removeChild(toast)
    }, 300)
  }, duration)
}

export function showSuccessToast(message: string, duration: number = 3000) {
  const toast = document.createElement('div')
  toast.className = 'success-toast'
  toast.textContent = message
  toast.style.cssText = `
    position: fixed;
    top: 20px;
    right: 20px;
    padding: 12px 20px;
    background: #67c23a;
    color: white;
    border-radius: 4px;
    box-shadow: 0 2px 12px rgba(0, 0, 0, 0.15);
    z-index: 9999;
    animation: slideIn 0.3s ease;
  `
  
  document.body.appendChild(toast)
  
  setTimeout(() => {
    toast.style.animation = 'slideOut 0.3s ease'
    setTimeout(() => {
      document.body.removeChild(toast)
    }, 300)
  }, duration)
}

// 添加动画样式
const style = document.createElement('style')
style.textContent = `
  @keyframes slideIn {
    from {
      transform: translateX(100%);
      opacity: 0;
    }
    to {
      transform: translateX(0);
      opacity: 1;
    }
  }
  
  @keyframes slideOut {
    from {
      transform: translateX(0);
      opacity: 1;
    }
    to {
      transform: translateX(100%);
      opacity: 0;
    }
  }
`
document.head.appendChild(style)

File: utils/error_handler.py
"""
全局异常处理模块
提供统一的错误处理和响应格式
"""

from typing import Dict, Any, Optional
from utils.logger import logger


class APIError(Exception):
    """API错误基类"""
    def __init__(self, message: str, status_code: int = 500, error_code: str = 'INTERNAL_ERROR'):
        self.message = message
        self.status_code = status_code
        self.error_code = error_code
        super().__init__(self.message)


class ValidationError(APIError):
    """验证错误"""
    def __init__(self, message: str):
        super().__init__(message, 400, 'VALIDATION_ERROR')


class NotFoundError(APIError):
    """资源未找到错误"""
    def __init__(self, message: str = '资源未找到'):
        super().__init__(message, 404, 'NOT_FOUND')


class ConflictError(APIError):
    """冲突错误"""
    def __init__(self, message: str):
        super().__init__(message, 409, 'CONFLICT')


class UnauthorizedError(APIError):
    """未授权错误"""
    def __init__(self, message: str = '未授权'):
        super().__init__(message, 401, 'UNAUTHORIZED')


class ForbiddenError(APIError):
    """禁止访问错误"""
    def __init__(self, message: str = '禁止访问'):
        super().__init__(message, 403, 'FORBIDDEN')


class RateLimitError(APIError):
    """限流错误"""
    def __init__(self, message: str = '请求过于频繁'):
        super().__init__(message, 429, 'RATE_LIMIT_EXCEEDED')


class ServiceUnavailableError(APIError):
    """服务不可用错误"""
    def __init__(self, message: str = '服务暂时不可用'):
        super().__init__(message, 503, 'SERVICE_UNAVAILABLE')


def handle_api_error(error: Exception) -> Dict[str, Any]:
    """
    处理API错误，返回统一的错误响应格式
    
    Args:
        error: 异常对象
        
    Returns:
        标准化的错误响应字典
    """
    if isinstance(error, APIError):
        # 已知的API错误
        logger.warning(f"API错误: {error.error_code} - {error.message}")
        return {
            'success': False,
            'error': {
                'code': error.error_code,
                'message': error.message,
                'status_code': error.status_code
            }
        }
    else:
        # 未知错误
        logger.error(f"未处理的异常: {type(error).__name__} - {str(error)}", exc_info=True)
        return {
            'success': False,
            'error': {
                'code': 'INTERNAL_ERROR',
                'message': '服务器内部错误',
                'status_code': 500
            }
        }


def create_error_response(error: Exception) -> tuple[Dict[str, Any], int]:
    """
    创建错误HTTP响应
    
    Args:
        error: 异常对象
        
    Returns:
        (响应字典, HTTP状态码)
    """
    error_data = handle_api_error(error)
    status_code = error_data['error']['status_code']
    return error_data, status_code


def log_request(method: str, path: str, status_code: int, duration: float, user_agent: str = ''):
    """
    记录API请求日志
    
    Args:
        method: HTTP方法
        path: 请求路径
        status_code: 响应状态码
        duration: 请求处理时间（秒）
        user_agent: 用户代理字符串
    """
    log_level = 'info' if status_code < 400 else 'warning' if status_code < 500 else 'error'
    
    log_message = f"{method} {path} - {status_code} - {duration:.3f}s"
    if user_agent:
        log_message += f" - {user_agent}"
    
    getattr(logger, log_level)(log_message)